Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -1047,9 +1049,134 @@ public class TypeArgumentInstantiation {
     protected abstract ILvalue EmitMemberSelect(Action<ConcreteSyntaxTree> obj, Type objType, MemberDecl member, List<TypeArgumentInstantiation> typeArgs, Dictionary<TypeParameter, Type> typeMap,
       Type expectedType, string/*?*/ additionalCustomParameter = null, bool internalAccess = false);
 
-    protected void EmitArraySelect(string index, Type elmtType, ConcreteSyntaxTree wr) {
-      EmitArraySelect(new List<string>() { index }, elmtType, wr);
+    protected enum MemberCompileStatus { Ordinary, Identity, AlwaysTrue }
+
+    protected MemberCompileStatus GetMemberStatus(MemberDecl member) {
+      if (member.EnclosingClass is DatatypeDecl dt) {
+        if (OptimizesInvisibleDatatypeWrappers && IsInvisibleWrapper(dt, out var dtor) && dtor == member) {
+          // "member" is the sole destructor of an invisible datatype wrapper
+          return MemberCompileStatus.Identity;
+        } else if (member is DatatypeDiscriminator) {
+          // a discriminator of an inductive or coinductive datatype
+          return dt.Ctors.Count(c => !c.IsGhost) == 1 ? MemberCompileStatus.AlwaysTrue : MemberCompileStatus.Ordinary;
+        }
+      }
+
+      return MemberCompileStatus.Ordinary;
     }
+
+    protected bool IsInvisibleWrapper(DatatypeDecl dt, out DatatypeDestructor coreDestructor) {
+      if (!DafnyOptions.O.OptimizeInvisibleDatatypeWrappers ||
+          !OptimizesInvisibleDatatypeWrappers ||
+          !(dt is IndDatatypeDecl) ||
+          dt.IsExtern(out _, out _) ||
+          dt.Members.Any(member => member is Field)) {
+        // don't optimize
+      } else {
+        var i = dt.Ctors.FindIndex(ctor => !ctor.IsGhost);
+        if (0 <= i) {
+          if (dt.Ctors.FindIndex(i + 1, ctor => !ctor.IsGhost) == -1) {
+            // there is exactly one non-ghost constructor
+            var ctor = dt.Ctors[i];
+            var j = ctor.Destructors.FindIndex(dtor => !dtor.IsGhost);
+            if (0 <= j) {
+              if (ctor.Destructors.FindIndex(j + 1, dtor => !dtor.IsGhost) == -1) {
+                // there is exactly one non-ghost parameter to "ctor"
+                var candidateCoreDestructor = ctor.Destructors[j];
+                var properlyContainedTypeDecls = new HashSet<TopLevelDecl>();
+                CompiledTypes(candidateCoreDestructor.Type, properlyContainedTypeDecls);
+                if (!properlyContainedTypeDecls.Contains(dt)) {
```

Comment:
I'm not sure I understand the purpose of this check. Why can we not always optimise away the constructor at this point?

If I have

```dafny
datatype Wrapper = Wrapper(ints: Ints)
datatype Ints = Ints(x: int, y: int) | AnotherWrapper(wrapper: Wrapper)
```

where it seems `properlyContainedTypeDecls.Contains(dt) == true`, then wouldn't we still be able to write that as

```dafny
datatype Ints = Ints(x: int, y: int) | AnotherWrapper(wrapper: Ints)
```

?

Reply:
Very good catch! The current algorithm is too strict and misses the opportunity to optimize your `Wrapper` type. I'll improve the algorithm.

As for the question about the purpose of the check, here is an answer: If optimizing away the constructor gives back the type itself, then there would be a problem. To illustrate, it may help to imagine a type like

``` dafny
datatype Cycle = Ctor(Cycle)
```

It would not be possible to remove the wrapper `Ctor`, because doing so would give another type whose wrapper `Ctor` we'd remove, and it is also a wrapper `Ctor` that we'd remove, etc. However, Dafny doesn't actually allow an `datatype` like `Cycle`, so to illustrate the real problem, I have to show a more complicated type. Here is one:

``` dafny
datatype GList =
  | ghost End
  | More(GList)
```

Dafny does allow this type. We don't represent `End` at run time, since it's ghost. The other constructor has one argument, so it looks like it could be optimized. But this would then run into the infinite-cycle problem I gave a taste of with the `Cycle` type above.